lsp/nls/src/background.rs

```diff
@@ -0,0 +1,396 @@
+//   fails.
+// - We ignore missing field errors. It would be nice not to ignore them, but it's hard
+//   to tell when they're appropriate: the term might intentionally be a partial configuration.
+fn eval_permissive(
+    vm: &mut VirtualMachine<impl ImportResolver, impl Cache>,
+    errors: &mut Vec<EvalError>,
+    rt: RichTerm,
+) {
+    match vm.eval(rt) {
+        Err(e) => errors.push(e),
+        Ok(t) => match t.as_ref() {
+            Term::Array(ts, _) => {
+                for t in ts.iter() {
+                    // After eval_closure, all the array elements  are
+                    // closurized already, so we don't need to do any tracking
+                    // of the env.
+                    eval_permissive(vm, errors, t.clone());
+                }
+            }
+            Term::Record(data) => {
+                for field in data.fields.values() {
+                    if let Some(v) = &field.value {
+                        let value_with_ctr = RuntimeContract::apply_all(
+                            v.clone(),
+                            field.pending_contracts.iter().cloned(),
+                            v.pos,
+                        );
+                        eval_permissive(vm, errors, value_with_ctr);
+                    }
+                }
+            }
+            _ => {}
+        },
+    }
+}
+
+// Returning an Option, just to let us use `?` when channels disconnect.
+fn worker(cmd_rx: IpcReceiver<Command>, response_tx: IpcSender<Response>) -> Option<()> {
+    let mut evals = Vec::new();
+    let mut cmds = Vec::new();
+    let mut world = World::default();
+
+    loop {
+        for cmd in cmds.drain(..) {
+            // Process all the file updates first, even if it's out of order with the evals.
+            // (Is there any use case for wanting the eval before updating the contents?)
+            match cmd {
+                Command::UpdateFile { uri, text } => {
+                    // Failing to update a file's contents is bad, so we terminate (and restart)
+                    // the worker.
+                    world.update_file(uri, text).unwrap();
+                }
+                Command::EvalFile { uri } => {
+                    evals.push(uri);
+                }
+            }
+        }
+
+        // Deduplicate the evals back-to-front, and then evaluate front-to-back.
+        // This means that if we get requests for `foo.ncl` and `bar.ncl`, and
+        // then another request for `foo.ncl` before started working on the
+        // first one, then we'll throw away the first `foo.ncl` and evaluate
+        // `bar.ncl` first.
+        let mut seen = HashSet::new();
+        let mut dedup = Vec::new();
+        for path in evals.iter().rev() {
+            if seen.insert(path) {
+                dedup.push(path.clone());
+            }
+        }
+        evals = dedup;
+
+        if let Some(uri) = evals.pop() {
```

Comment:
What happens here if we have, say, 10 evals inside `evals`? It seems to me that we're going to pop one, and then do another round of loop. Say we receive new commands, so we recompute dedup. Then, L147 (two lines above), we assign `evals` to `dedup`. Won't this erase the `9` remaining evals that were saved from the previous loop? Shouldn't we prepopulate `dedup` with the evals from the previous round or something?

Reply:
Sorry, you can ignore that remark. We iterate on evals, I misunderstood the code above.